src/test/pythonEnvironments/base/locators/composite/environmentsResolver.unit.test.ts

```diff
@@ -152,12 +160,43 @@ suite('Python envs locator - Environments Resolver', () => {
         });
     });
 
+    test('No updates except the null event is sent if environment is not safe to execute', async () => {
+        // Arrange
+        isEnvSafe.returns(false);
+        const env1 = createNamedEnv('env1', '3.5.12b1', PythonEnvKind.Unknown, path.join('path', 'to', 'exec1'));
+        const env2 = createNamedEnv('env2', '3.8.1', PythonEnvKind.Unknown, path.join('path', 'to', 'exec2'));
+        const environmentsToBeIterated = [env1, env2];
+        const parentLocator = new SimpleLocator(environmentsToBeIterated);
+        const onUpdatedEvents: (PythonEnvUpdatedEvent | null)[] = [];
+        const resolver = new PythonEnvsResolver(parentLocator, envInfoService, environmentsSecurity);
+
+        const iterator = resolver.iterEnvs(); // Act
+
+        // Assert
+        let { onUpdated } = iterator;
+        expect(onUpdated).to.not.equal(undefined, '');
+
+        // Arrange
+        onUpdated = onUpdated!;
+        onUpdated((e) => {
+            onUpdatedEvents.push(e);
+        });
+
+        // Act
+        await getEnvs(iterator);
+        await sleep(1); // Resolve pending calls in the background

```

Comment:
Rather than sleeping, you can be explicit about it:

```suggestion
        const ready = createDeferred<void>();
        onUpdated((e) => {
            onUpdatedEvents.push(e);
            if (e === null) {
                ready.resolve();
            }
        });

        // Act
        await getEnvs(iterator);
        await ready.promise; // Resolve pending calls in the background
```

Reply:
I just found the older code more cleaner, but this will do.